### PR TITLE
Train a machine learning model how-to

### PR DESCRIPTION
**Pull request by luisquintanilla**
Train a machine learning model how-to

[Internal Review Link](https://review.docs.microsoft.com/en-us/dotnet/machine-learning/how-to-guides/train-machine-learning-model-ml-net?branch=pr-en-us-11771)

[AB#1504318](https://dev.azure.com/mseng/677da0fb-b067-4f77-b89b-f32c12bb8617/_workitems/edit/1504318)

### CODE REVIEW

File: docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+ms.date: 04/09/2019
+author: luisquintanilla
+ms.author: luquinta
+ms.custom: mvc,how-to
+#Customer intent: As a developer I want to know how to train my model
+---
+
+# How-To: Learn how to train a machine learning model in ML.NET
+
+This how-to shows you how to build machine learning models, extract learned parameters and measure performance with ML.NET. Although this sample goes through the training of a regression model, the concepts are applicable throughout a majority of the other algorithms.
```
Comment (JRAlexander) [inline docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: Would this work? "Although this sample trains a regression model, the concepts are applicable throughout a majority of the other algorithms."

File: docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    new HousingData
+    {
+        Size = 550f,
+        HistoricalPrices = new float[] { 99000f, 98000f, 130000f },
+        CurrentPrice = 180000f
+    }
+};
+```
+
+Use the [`TrainTestSplit`](xref:Microsoft.ML.DataOperationsCatalog.TrainTestSplit*) method to split the data into train and test sets. The result will be a [`TrainTestData`](xref:Microsoft.ML.DataOperationsCatalog.TrainTestData) object which contains two [`IDataView`](xref:Microsoft.ML.IDataView) members, one for the train set and the other for the test set. How much of the data is split is determined by the `testFraction` parameter. The snippet below is holding out 20 percent of the original data for the test set.
```
Comment (JRAlexander) [inline docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: Would this work? "The data split percentage is determined by the `testFraction` parameter."

File: docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+```csharp
+DataOperationsCatalog.TrainTestData dataSplit = mlContext.Data.TrainTestSplit(data, testFraction: 0.2);
+IDataView trainData = dataSplit.TrainSet;
+IDataView testData = dataSplit.TestSet;
+```
+
+## Preapre the data
```
Comment (JRAlexander) [inline docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: "Preapre" -> "Prepare"

File: docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+Before using the data to train a machine learning model, it needs to be pre-processed. More information on data preparation can be found on the [data prep how-to article](prepare-data-ml-net.md) as well as the [`transforms page`](../resources/transforms.md).
```
Comment (JRAlexander) [inline docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: How about something like: "The data needs to be pre-processed before training a machine learning model."

File: docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+ML.NET applies certain defaults with regards to column types and names when building machine learning models .
+
+### Working with default column types
+
+By default, the input expected by the machine learning algorithms is a numerical vector. One way to specify that the input data will be a vector is in the data model with the [`VectorType`](xref:Microsoft.ML.Data.VectorTypeAttribute) attribute. This approach is appropriate when all of the data is already in numerical format and is intended to be ingested and processed together (i.e. image pixels).
```
Comment (JRAlexander) [inline docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: How about something like: "The machine learning algorithms default input  is a numerical vector. Use the [`VectorType`](xref:Microsoft.ML.Data.VectorTypeAttribute) attribute for vector input data."?

File: docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+Additionally, algorithms use default column names when none are specified. All trainers have a parameter called `featureColumnName` for the inputs of the algorithm and when applicable they also have a parameter for the expected value called `labelColumnName`. By default those values are `Features` and `Label` respectively. 
+
+By using the [`Concatenate`](xref:Microsoft.ML.TransformExtensionsCatalog.Concatenate*) method during pre-processing to create a new column called `Features`, there is no need to specify the feature column name in the parameters of the algorithm since it already exists in the pre-processed `IDataView`. The label column is `CurrentPrice`, but since the [`ColumnName`](xref:Microsoft.ML.Data.ColumnNameAttribute.#ctor*) attribute is used in the data model, ML.NET renames the `CurrentPrice` column to `Label` which removes the need to provide the `labelColumnName` parameter to the machine learning algorithm estimator. 
+
+If you do not want to use the default column names, pass in the names of the feature and label columns as parameters when defining the machine learning algorithm estimator as demonstrated by the subsequent snippet:
```
Comment (JRAlexander) [inline docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: Don't

File: docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+#Customer intent: As a developer I want to know how to train my model
+---
+
+# How-To: Learn how to train a machine learning model in ML.NET
+
+This how-to shows you how to build machine learning models, extract learned parameters and measure performance with ML.NET. Although this sample goes through the training of a regression model, the concepts are applicable throughout a majority of the other algorithms.
+
+## Split data for training and testing
+
+The goal of a good machine learning model is to generalize well or predict unseen observations with more accuracy than those it was trained on. To do so, it is recommended to split data into training and test sets.
```
Comment (JRAlexander) [inline docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: those = the initial observations?  "it is" -> "it's" Could we be more prescriptive and use something like:  "To do so, split data into training and test sets. "?

**Review by JRAlexander — CHANGES_REQUESTED** [review]
Just a few changes.

File: docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+DataOperationsCatalog.TrainTestData dataSplit = mlContext.Data.TrainTestSplit(data, testFraction: 0.2);
+IDataView trainData = dataSplit.TrainSet;
+IDataView testData = dataSplit.TestSet;
+```
+
+## Prepare the data
+
+The data needs to be pre-processed before training a machine learning model. More information on data preparation can be found on the [data prep how-to article](prepare-data-ml-net.md) as well as the [`transforms page`](../resources/transforms.md).
+
+ML.NET applies certain defaults with regards to column types and names when building machine learning models .
```
Comment (natke) [inline docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: Suggest: "In ML.NET, data is represented by <xref:Microsoft.ML.IDataView> objects. A data view is made up of rows and columns, where each column has a name, type and size, and each row is an instance of all of the columns that make up the data view. Columns have default names and types.

File: docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+// Build machine learning model
+var trainedModel = sdcaEstimator.Fit(transformedTrainingData);
+```
+
+## Extract model parameters
+
+After the model has been trained, extract the learned [`ModelParameters`](xref:Microsoft.ML.Trainers.ModelParametersBase`1) for inspection or re-training. The [`LinearRegressionModelParameters`](xref:Microsoft.ML.Trainers.LinearRegressionModelParameters) provide the bias and learned coefficients or weights of the trained model. 
+
+```csharp
+var trainedModelParameters = trainedModel.Model as LinearRegressionModelParameters;
```
Comment (natke) [inline docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: Do we want to clarify that this only works if the algorithm is the last transformation in the pipeline?

**Review by natke — CHANGES_REQUESTED** [review]
Looks great! I've added a few comments.

File: docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md (hunk 9 above)
Comment (luisquintanilla) [reply docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: In this case as well as the other samples I split out the pipelines so that you should always be able to extract the model parameters.

File: docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+#Customer intent: As a developer I want to know how to train and evaluate a machine learning model with ML.NET
+---
+
+# How-To: Learn how to train and evaluate a machine learning model in ML.NET
+
+This how-to shows you how to build machine learning models, extract learned parameters and measure performance with ML.NET. Although this sample trains a regression model, the concepts are applicable throughout a majority of the other algorithms.
+
+## Split data for training and testing sets
+
+The goal of a good machine learning model is to generalize well or predict unseen observations with more accuracy than the initial observations it was trained on. To do so, split data into training and test sets.
```
[the remark this answers is not in the document]
Comment (luisquintanilla) [reply docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: I can change it to "better". If you're getting better results on your train than on test data chances are the model is overfitting.

[the remark this answers is not in the document]
Comment (luisquintanilla) [reply docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: Also accuracy can get confused with the actual metric so I can change accuracy to performance.

File: docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md (hunk 8 above)
Comment (luisquintanilla) [reply docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: I'm not sure if this fits because the columns don't have default names and types. These default names and types refer to those expected by the ML algo estimator.

File: docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+#Customer intent: As a developer I want to know how to train and evaluate a machine learning model with ML.NET
+---
+
+# How-To: Learn how to train and evaluate a machine learning model in ML.NET
+
+Learn how to build machine learning models, extract learned parameters and measure performance with ML.NET. Although this sample trains a regression model, the concepts are applicable throughout a majority of the other algorithms.
+
+## Split data for training and testing sets
+
+The goal of a good machine learning model is to generalize well or predict better on unseen observations than the initial observations it was trained on. To do so, split data into training and test sets.
```
Comment (sfilipi) [inline docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: >better [](start = 75, length = 6)

I'd remove better, because it seems like there are already predictions, and we're trying to make them better.

File: docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+---
+
+# How-To: Learn how to train and evaluate a machine learning model in ML.NET
+
+Learn how to build machine learning models, extract learned parameters and measure performance with ML.NET. Although this sample trains a regression model, the concepts are applicable throughout a majority of the other algorithms.
+
+## Split data for training and testing sets
+
+The goal of a good machine learning model is to generalize well or predict better on unseen observations than the initial observations it was trained on. To do so, split data into training and test sets. 
+
```
Comment (sfilipi) [inline docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: Maybe: The goal of a good supervised machine learning model is to generalize the relationship between the known variables (features), and one unknown related variable (label) and use the learnt pattern to predict the label value for a new set of features.

File: docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+    [VectorType(3)]
+    public float[] HistoricalPrices { get; set; }
+
+    [LoadColumn(4)]
+    [ColumnName("Label")]
+    public float CurrentPrice { get; set; }
+}
+```
+
+Load the data into an [`IDataView`](xref:Microsoft.ML.IDataView):
```
Comment (sfilipi) [inline docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: >[`IDataView`](xr [](start = 22, length = 16)

should it add: the ML.Net data format?

File: docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+## Prepare the data
+
+The data needs to be pre-processed before training a machine learning model. More information on data preparation can be found on the [data prep how-to article](prepare-data-ml-net.md) as well as the [`transforms page`](../resources/transforms.md).
+
+ML.NET algorithms have constraints on input column types. Additionally, default values are used for input and output column names when no values are specified.
+
+### Working with expected column types
+
+The machine learning algorithms in ML.NET expect a numerical vector as input. Apply the [`VectorType`](xref:Microsoft.ML.Data.VectorTypeAttribute) attribute to your data model when all of the data is already in numerical format and is intended to be processed together (i.e. image pixels).
```
Comment (sfilipi) [inline docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: >a numerical vector [](start = 49, length = 18)

a vector of floats with known size.

File: docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+## Extract model parameters
+
+After the model has been trained, extract the learned [`ModelParameters`](xref:Microsoft.ML.Trainers.ModelParametersBase`1) for inspection or re-training. The [`LinearRegressionModelParameters`](xref:Microsoft.ML.Trainers.LinearRegressionModelParameters) provide the bias and learned coefficients or weights of the trained model. 
+
+```csharp
+var trainedModelParameters = trainedModel.Model as LinearRegressionModelParameters;
+```
+
+> [!NOTE]
+> Other models have parameters that are specific to their tasks. For example, K-Means has centroids and the [`KMeansModelParameters`](xref:Microsoft.ML.Trainers.KMeansModelParameters) contains a property that stores these learned centroids. To learn more, visit the [`Microsoft.ML.Trainers` API Documentation](xref:Microsoft.ML.Trainers) and look for classes that contain `ModelParameters` in their name.
```
Comment (sfilipi) [inline docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: >K-Means [](start = 78, length = 7)

provide a wikipedia link? or documentation link?

File: docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+```csharp
+var trainedModelParameters = trainedModel.Model as LinearRegressionModelParameters;
+```
+
+> [!NOTE]
+> Other models have parameters that are specific to their tasks. For example, K-Means has centroids and the [`KMeansModelParameters`](xref:Microsoft.ML.Trainers.KMeansModelParameters) contains a property that stores these learned centroids. To learn more, visit the [`Microsoft.ML.Trainers` API Documentation](xref:Microsoft.ML.Trainers) and look for classes that contain `ModelParameters` in their name. 
+
+## Evaluate model quality
+
+To help choose the best performing model, it is essential to evaluate its performance on unseen data. Use the [`Evaluate`](xref:Microsoft.ML.RegressionCatalog.Evaluate*) method, to measure calculate various metrics for the trained model.
```
Comment (sfilipi) [inline docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: >unseen [](start = 89, length = 6)

test data (because they have to have a label.)
unseen data is mostly used for prediction.

Comment (sfilipi) [inline docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: >measure calculate [](start = 181, length = 17)

imo either measure or calculate

**Review by sfilipi — CHANGES_REQUESTED** [review]
:clock1:

File: docs/machine-learning/toc.yml [hunk 17]
```diff
@@ -35,8 +35,8 @@
       href: how-to-guides/train-cross-validation-ml-net.md
     - name: Inspect intermediate pipeline data values
       href: how-to-guides/inspect-intermediate-data-ml-net.md
-    - name: Train regression model
-      href: how-to-guides/train-regression-model-ml-net.md
+    - name: Train a model
+      href: how-to-guides/train-machine-learning-model-ml-net.md
```
Comment (natke) [inline docs/machine-learning/toc.yml]: Can we move this to the top of this section of the TOC?

File: docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+author: luisquintanilla
+ms.author: luquinta
+ms.topic: how-to
+ms.custom: mvc
+#Customer intent: As a developer I want to know how to train and evaluate a machine learning model with ML.NET
+---
+
+# How-To: Learn how to train and evaluate a machine learning model in ML.NET
+
+Learn how to build machine learning models, extract learned parameters and measure performance with ML.NET. Although this sample trains a regression model, the concepts are applicable throughout a majority of the other algorithms.
```
Comment (natke) [inline docs/machine-learning/how-to-guides/train-machine-learning-model-ml-net.md]: Can we do a line break before the Although?

**Review by natke — APPROVED** [review]

**Review by JRAlexander — APPROVED** [review]
